Description: Fix support pty on GNU/Hurd
Author: Jonas Smedegaard <dr@jones.dk>
Last-Update: 2014-08-20

--- a/plugins/forkptyrouter/forkptyrouter.c
+++ b/plugins/forkptyrouter/forkptyrouter.c
@@ -16,7 +16,7 @@
 
 extern struct uwsgi_server uwsgi;
 
-#if defined(__linux__) || defined(__GNU_kFreeBSD__)
+#if defined(__linux__) || defined(__GNU_kFreeBSD__) || defined(__HURD__)
 #include <pty.h>
 #elif defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
 #include <util.h>
--- a/plugins/pty/pty.c
+++ b/plugins/pty/pty.c
@@ -1,5 +1,5 @@
 #include <uwsgi.h>
-#if defined(__linux__) || defined(__GNU_kFreeBSD__)
+#if defined(__linux__) || defined(__GNU_kFreeBSD__)) || defined(__HURD__)
 #include <pty.h>
 #elif defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
 #include <util.h>
